GGRC-1950 Move migrations to appengine

sskhatri opened the pull request:
This PR also enables maintenance mode while migrations are running.

Smotko commented on bin/do_deploy:
```diff
@@ -49,8 +49,13 @@ write_visible_message 'Authenticating'
-yes | gcloud app deploy src/cron.yaml
-yes | gcloud app deploy src/queue.yaml
+
+if [-z "${AUTHORIZED_DOMAINS}" ]
```
We've had issues with cron/queue not being setup correctly in the past. @zidarsk8 is this really a good solution to this problem?

sskhatri replied:
I don't think this is a solution at all. I am just doing what is currently being done. This problem is with user facing instances only and Akhil mentioned that he comment these two lines while deploying to bypass this problem.

Smotko replied:
I know, but putting this behind an if clause doesn't really address the problem that we are having.

Smotko commented on src/ggrc/app.py:
```diff
@@ -66,6 +72,25 @@ def setup_user_timezone_offset():
+@app.before_request
+def check_if_under_maintenance():
+  """Check if the site is in maintenance mode."""
+  from ggrc.models.maintenance import Maintenance
```
Can we wrap this up in our `with benchmark('Check for maintenance'):` so that we can track how much time it takes to do the check.

sskhatri replied:
Done.

Smotko commented on src/ggrc/migrate.py:
```diff
@@ -96,11 +105,39 @@ def all_extensions():
+      
+    # For testing purpose. Remove two lines of code below.
+    import time
+    time.sleep(60)
```
I believe this needs to be removed.

sskhatri replied:
Done.

Smotko commented on src/ggrc/migrations/versions/20170922170856_3d8df1f399a0_add_maintenance_table.py:
```diff
@@ -0,0 +1,41 @@
+
+# revision identifiers, used by Alembic.
+revision = '3d8df1f399a0'
+down_revision = '434683ceff87'
```
The down revision needs to be updated.

sskhatri replied:
Done.

Smotko commented on bin/do_deploy:
```diff
@@ -54,3 +52,8 @@ write_visible_message "Deploying src/app.yaml into version $VERSION"
+
+if [ -z "${AUTHORIZED_DOMAINS}" ]
+then
+  curl --data "access_token=${ACCESS_TOKEN}" https://${VERSION}-dot-${APPENGINE_INSTANCE}.appspot.com/maintenance/migrate
```
We need to let the deployment script know if the migration was successful because it needs to fail if something goes wrong.

sskhatri replied:
Deployment script now fails if there is issue with migration.

Smotko commented on src/ggrc/maintenance_views/maintenance.py:
```diff
@@ -77,7 +84,7 @@ def authenticate():
   else:
     gae_user = users.get_current_user()
-    if gae_user and gae_user.email() in settings.BOOTSTRAP_ADMIN_USERS:
+    if gae_user and gae_user.email() in settings.BOOTSTRAP_ADMIN_USERS or gae_user.email() in ['skhatri@google.com']:
```
No google emails should be saved in the open source repo.

sskhatri replied:
Good catch. It was added for testing purpose but forgot to remove.
Done.

Smotko commented on src/ggrc/app.py:
```diff
@@ -62,11 +68,35 @@ def setup_user_timezone_offset():
+  from ggrc.models.maintenance import Maintenance
+  sess = db.session
+  try:
+    db_row = sess.query(Maintenance).get(1)
```
An idea:

What if the migration script would insert a row into the maintenance table on every run? I think it would be beneficial to see the log of migration runs + we could store things like the error log that might help with the issue I pointed out here: https://github.com/google/ggrc-core/pull/6489#pullrequestreview-68905717

sskhatri replied:
Separate migration_log table created for storing history of migration runs, its status and logs.

Smotko commented on src/ggrc/migrations/versions/20170922170856_3d8df1f399a0_add_maintenance_table.py:
```diff
@@ -0,0 +1,37 @@
+
+# revision identifiers, used by Alembic.
+revision = '3d8df1f399a0'
+down_revision = '434683ceff87'
```
Needs to be updated.

sskhatri replied:
Done.

Smotko commented on src/ggrc/maintenance_views/maintenance.py:
```diff
@@ -0,0 +1,151 @@
+      else:
+        msg = "Invalid access token"
+        logger.info(msg)
+        return msg
```
We should never return response code 200, but an error like "Invalid access token". This needs to run 403 Forbidden

sskhatri replied:
Done

Smotko commented on src/ggrc/maintenance_views/maintenance.py:
```diff
@@ -0,0 +1,151 @@
+    else:
+      msg = "Access token not found in settings"
+      logger.info(msg)
+      return msg
```
Again this shouldn't return 2xx, it should return a server error 5xx

sskhatri replied:
Done

Smotko commented on src/ggrc/maintenance_views/maintenance.py:
```diff
@@ -0,0 +1,151 @@
+    else:
+      msg = "User not authorized"
+      logger.info(msg)
+      return msg
```
Again 403 should have been returned.

sskhatri replied:
Done

Smotko commented on src/ggrc/maintenance_views/maintenance.py:
```diff
@@ -0,0 +1,151 @@
+    if hasattr(settings, 'ACCESS_TOKEN'):
+      if request.form.get("access_token") == settings.ACCESS_TOKEN:
+        mig_row_id = run_migration()
+        return "{0},{1},{2}".format('Migration is running in background.',
```
We shouldn't try to avoid custom formats like this as much as possible. 

Because this request initialized a background task, we should return 202 Accepted status and a json string with the id of the background task + a message if needed.

sskhatri replied:
Done. Please take a look at updated code.

Smotko commented on src/ggrc/maintenance_views/maintenance.py:
```diff
@@ -0,0 +1,151 @@
+        logger.info(msg)
+        return msg
+    else:
+      msg = "Access token not found in settings"
```
We should never disclose server configuration to an unauthorized user. It's okay to store this message into the logs, but this message shouldn't be sent back to the user.

sskhatri replied:
Done

Smotko commented on src/ggrc/maintenance_views/maintenance.py:
```diff
@@ -0,0 +1,151 @@
+  """Authenticates user and allows to run migration."""
+  if "access_token" in request.form:
+    if hasattr(settings, 'ACCESS_TOKEN'):
+      if request.form.get("access_token") == settings.ACCESS_TOKEN:
```
These nested if clauses would be much easier to read if they were inverted so that all the nesting would be avoided e.g.:

```
if "access_token" not in requst_form:
  return "Missing access_token parameter", 500

if not (hasattr(settings, 'ACCESS_TOKEN') and request.form.get("access_token") == settings.ACCESS_TOKEN):
  return "Unauthorized", 403

# Run migration code here
return response_json, 202
```

sskhatri replied:
I have updated the code as per your suggestion.

Smotko commented on src/ggrc/maintenance_views/maintenance.py:
```diff
@@ -0,0 +1,151 @@
+      return "Fail,{}".format("No migration entry in db.")
+  except sqlalchemy.exc.ProgrammingError as e:
+    if not re.search(r"""\(1146, "Table '.+' doesn't exist"\)$""", e.message):
+      raise
```
Is this code just suppressing all table doesn't  exist errors? Shouldn't we at at least log it when that happens?

sskhatri replied:
Actually it suppresses these errors because when migration is run for the first time, the maintenance tables are also part of migration and it won't exist for the first time.

But you are right we should at least log it because it might occur due to other table not existing.

Smotko commented on src/ggrc/maintenance_views/maintenance.py:
```diff
@@ -0,0 +1,151 @@
+    sess = db.session
+    maint_row = sess.query(Maintenance).get(1)
+    mig_row = sess.query(MigrationLog).get(row_id)
+    if mig_row:
```
I feel like these if clauses are extremely hard to follow, can you try to rewrite them similar to the way I suggested above? Also instead of comma separated list I think we should always be returning json if the API is going to be consumed by a script or a template if it's going to be displayed as html.

sskhatri replied:
I have tried to rewrite them in the similar way you suggested.

Smotko commented on bin/do_deploy:
```diff
@@ -42,15 +42,66 @@ source "$SETTINGS_FILE"
+  	    break
+  	  elif [ "${status_msg[0]}" == "In progress" ]
+  	  then
+  	    write_visible_message "Migration is in progress. Retrying in 1 minute."
```
Most migrations should be finished in a few seconds, waiting for 1 minute before retries should be too much I think, but it would be cool if the wait time went up after every request (first req waits for 1s next one 2s then 4s then 8s ...)

Smotko commented on bin/do_deploy:
```diff
@@ -42,15 +42,66 @@ source "$SETTINGS_FILE"
+COUNTER=0
+if [ -z "${AUTHORIZED_DOMAINS}" ]
+then
+  response=$(curl --data "access_token=${ACCESS_TOKEN}" https://${VERSION}-dot-${APPENGINE_INSTANCE}.appspot.com/maintenance/migrate)
```
Is this always going to be appspot.com? Even for googleplex instances?

sskhatri replied:
Updated the code to look for APPENGINE_INSTANCE variable instead of AUTHORIZED_DOMAINS.

The idea is to run migration automatically only for appspot instances because for googleplex instance it gives uberproxy error.

Smotko commented on bin/trigger_migrations.py:
```diff
@@ -0,0 +1,57 @@
+        break
+
+      if sleep_time > CUTOFF_TIME:
+        print "Migration running for too long. Aborting.."
```
The text is a bit misleading here. The migration will not actually be aborted, you'll just stop checking.

sskhatri replied:
Updated text

Smotko commented on src/ggrc/maintenance_views/maintenance.py:
```diff
@@ -0,0 +1,155 @@
+
+
+@maintenance_app.route('/maintenance/migrate', methods=['POST'])
+def authenticate():
```
I think we should also check if a migration is running here. Otherwise we risk running the migration twice. 

Also it might be a bit misleading that the function that handles `'/maintenance/migrate'` is named authenticate.

sskhatri replied:
Added a check in trigger_migration function.

Smotko commented on src/ggrc/maintenance_views/maintenance.py:
```diff
@@ -0,0 +1,155 @@
+def turn_off_maintenance_mode():
+  """Super users are allowed to turn off maintenance mode manually."""
+  gae_user = users.get_current_user()
+  if gae_user and gae_user.email() in settings.BOOTSTRAP_ADMIN_USERS:
```
Shouldn't we be able to turn off the maintenance mode with the `ACCESS_TOKEN`?

sskhatri replied:
Done.

sskhatri commented:
I authored this commit

Smotko commented on bin/trigger_migrations.py:
```diff
@@ -0,0 +1,57 @@
+  post_data = {'access_token': access_token}
+  response = requests.post(url, data=post_data)
+  response_body = json.loads(response.text)
+  print response_body.get("message")
```
Is this a forgotten print statement? We should use the logger if we want to log messages.

sskhatri replied:
Done

Smotko commented on bin/trigger_migrations.py:
```diff
@@ -0,0 +1,57 @@
+      response = requests.get(url)
+      response_body = json.loads(response.text)
+      if response.status_code != 202:
+        print "Unrecognized http response : {}, {}".format(
```
These sort of errors should be logged with logger.error and not a print statement.

sskhatri replied:
Done

zidarsk8 commented:
@sskhatri 
I have tried deploying with a broken migration and there were good notifications on that, and app still got deployed and went out of maintenance mode.

this was all I got:
```
  $ gcloud app browse
/vagrant-dev/opt/dev_virtualenv/local/lib/python2.7/site-packages/requests/__init__.py:80: RequestsDependencyWarning: urllib3 (1.22) or chardet (2.1.1) doesn't match a supported version!
  RequestsDependencyWarning)
Migration is in progress. Retrying in 1 seconds
Migration seem to stuck.
```

and app engine logs were no better
![capture 2017-11-03 at 19 22 12](https://user-images.githubusercontent.com/566311/32290323-05678fc6-bf3a-11e7-96fc-9285af9d50ca.png)

And the error message from deferred (or even a notification that there was an error) was not shown.

> migrations seem to stuck
does not look like a valid message when an actual error occurs, because "seems" applies we just think they are, but once maintenance mode is turned off, there is nothing that would make me think it did not work.

I believe that disabling the maintenance mode if something goes wrong is dangerous, because users can come in play with the app and break more stuff before we have time to fix it.

sskhatri commented:
Done. The code is updated not to turn off maintenance mode in case of failure. Also I have changed to better logging message.

BTW, the appengine logs in the screenshot above you shared actually has the useful error logs about the failure. We should be looking at /_ah/queue/deferred instead of /maintenance/migrate.

zidarsk8 commented:
removed "check migration chain" and added "kokoro:force-run" because migrations that were merged were in workflows module and do not affect the migration chain modified in this PR.

sskhatri commented:
@zidarsk8

I have tested deployment with my pull request and it went fine. As per your suggestion I tested following use cases:

1. Deployment on existing ggrc-perf-db
2. Deployment on fresh ggrc-perf-db

Also I have added a check to only allow BOOTSTRAP_ADMIN_USERS to visit and run maintenance dashboard.

There were couple of issues with my pull request which are resolved now.
1. While rebasing and resolving merge conflicts some files were removed from my db_migration2appengine branch.
2. When application is deployed, the appengine fires a first empty GET request (/_ah/start) and it was just redirecting when under_maintenance flag in maintenance table is set. This caused the connection not being released. I have added a logic to check if the request.path is '/_ah/start' in src/ggrc/app.py
